torch/distributed/_spmd/graph_optimization.py

```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+                assert step_args, step_args
+                orig_step_outputs.append(orig_optim_block.step.outputs[idx])
+            step = gm.graph.call_function(
+                torch.ops.aten._foreach_add.Scalar,
+                (step_args, 1),
+            )
+        step_block = ForeachAddBlock(step, generate_output=True)
+        for i, step_output in enumerate(step_block.outputs):
+            # Replace the original step output in the graph output node with
+            # the new one.
+            orig_step_output = orig_step_outputs[i]
+            replace_flatten_output_args(orig_step_output, step_output)
+            # Also need to replace the step output used for the new optimizer.
+            assert optim_args[group_idx].state_steps[i] == orig_step_output, (
+                f"The expected step output node mismatched, {orig_step_output} "
+                f"{optim_args[group_idx].state_steps[i]}"
+            )
+            optim_args[group_idx].state_steps[i] = step_output
+
+        # Insert the optimizer node after the first step output because its
+        # topo sort order is the last.
+        with gm.graph.inserting_after(step_block.outputs[0]):
+            optim = gm.graph.call_function(
+                torch.ops.aten._fused_adam.default,
+                optim_args[group_idx],
+                orig_optim_block.optim.optim_node.kwargs,
+            )
+        optim_block = FusedAdamBlock(optim, generate_output=True)
+        for curr_idx, orig_idx in enumerate(orig_optim_indices[group_idx]):
+            list_names = ("param_outputs", "exp_avgs_outputs", "exp_avg_sqs_outputs")
+            for name in list_names:
+                orig_list = getattr(orig_optim_block.optim, name)
+                curr_list = getattr(optim_block, name)
+                replace_flatten_output_args(orig_list[orig_idx], curr_list[curr_idx])
+
+        result.append(FusedOptimizerBlock(step_block, optim_block))
+
+    # Optimizer is used as the output of the train_step. Therefore, we have to
+    # update the output node of the graph.
+    output_args, output_kwargs = tree_unflatten(flatten_output_args, spec)
+    gm.graph.node_set_args(output, output_args)
+    gm.graph.node_set_kwargs(output, output_kwargs)
+    # Remove the original copy_ nodes as they won't be DCE.
+    for copy_output in itertools.chain(
+        orig_optim_block.optim.param_outputs,
+        orig_optim_block.optim.exp_avgs_outputs,
+        orig_optim_block.optim.exp_avg_sqs_outputs,
+    ):
+        gm.graph.erase_node(copy_output)
+    # Call DCE once to get rid of the old optimizer. By doing so, we will be
+    # able to erase the copy_ nodes of step later.
+    gm.graph.eliminate_dead_code()
+    for copy_output in orig_optim_block.step.outputs:
+        gm.graph.erase_node(copy_output)
+    # This is not required but calling this for consistency.
+    gm.graph.eliminate_dead_code()
+
+    return result[0], result[1]
+
+
+def split_fused_optimizer(
```

Comment:
hmm, this one doesn't seem to be plugged into `torch.distributed._spmd.compile`? I assume we need to provide a helper `gm_transformation`?

Reply:
how do we plan to construct `split_gradients`? Manually for now?

Reply:
nvm, I saw the pr on top is using this one